Adding a card of WPrime coupling only to 3rd generation in the process

sifuluo opened the pull request:
This card is made for the analysis that studies the pure 3rd generation coupling wprime.
The models can be found at
https://indico.cern.ch/event/1150625/contributions/4832490/attachments/2431728/4164399/AndrewAndSifu.pdf
On page 25 of the pdf or slide 6 of my part.


This model comes with two diagrams, that differ in either of the W bosons decays leptonically.
The mass points expected are 300GeV to 1100GeV with a step of 100GeV.
The copy.py scrip automatically generates cards of both diagrams and all mass points expected.

This model serves as a complimentary to Andrew's analysis, whose card is merged at:
https://github.com/cms-sw/genproductions/pull/2807/

The djr plots are show below as the first part of the file name indicates the xqcut and qcut parameter.

[10-15central_qCut_djr_plots.pdf](https://github.com/cms-sw/genproductions/files/8900599/10-15central_qCut_djr_plots.pdf)
[10-30central_qCut_djr_plots.pdf](https://github.com/cms-sw/genproductions/files/8900602/10-30central_qCut_djr_plots.pdf)
[30-45central_qCut_djr_plots.pdf](https://github.com/cms-sw/genproductions/files/8900603/30-45central_qCut_djr_plots.pdf)
[30-60central_qCut_djr_plots.pdf](https://github.com/cms-sw/genproductions/files/8900605/30-60central_qCut_djr_plots.pdf)
[60-90central_qCut_djr_plots.pdf](https://github.com/cms-sw/genproductions/files/8900611/60-90central_qCut_djr_plots.pdf)
[200-300central_qCut_djr_plots.pdf](https://github.com/cms-sw/genproductions/files/8900613/200-300central_qCut_djr_plots.pdf)

My process's final state comes with 3 b-jets and 2 light jets.
And one additional jets in one of the sub-processes.
In the djr plots, I only have 2 lines even if I explicitly set nJetMax = 5 in the pythia8_card when I run madgraph to produce these plots.
Is that expected or something I missed?
Also from the djr plots, I can't determine which one looks the most smooth to pick.

menglu21 commented on bin/MadGraph5_aMCatNLO/cards/production/13TeV/WPrime_tbOnly/WPrime_tbOnly_run_card.dat:
```diff
@@ -0,0 +1,270 @@
+# Apply pt/E/eta/dr/mij cuts on decay products or not
+# (note that etmiss/ptll/ptheavy/ht/sorted cuts always apply)
+#**********************************************************
+   F  = cut_decays    ! Cut decay products
```
do you want to keep "cut_decays" consistent with PR2807?

sifuluo replied:
Yes, thank you for noticing that.

menglu21 commented:
As for the jet matching setup, you should set njetmax=1 since there is only 1 QCD jet which needs matching. for the djr, you are expected to check plots like https://deschwar.web.cern.ch/deschwar/tWZ/plots/DJR/DJR01_WZ.pdf to determine the xqcut and qcut

sihyunjeon commented on bin/MadGraph5_aMCatNLO/cards/production/13TeV/WPrime_tbOnly/WPrime_tbOnly_FL_proc_card.dat:
```diff
@@ -0,0 +1,15 @@
+define ts = t t~
+generate p p > ts wp, (ts > bs ws, ws > ls ls ),(wp > bs ts, (ts > bs ws, ws > j j))
+add process p p > ts wp j, (ts > bs ws, ws > ls ls ),(wp > bs ts, (ts > bs ws, ws > j j))
+add process p p > ts wp, (ts > bs ws j, ws > ls ls ),(wp > bs ts, (ts > bs ws, ws > j j))
```
chiming in a bit `ts > bs ws j` this is also not possible

sifuluo replied:
You are right about the overlap.
![L10p1](https://user-images.githubusercontent.com/18496702/174639726-60d4dc00-635c-401e-a188-9a8680e2aa61.png)
![L10p2](https://user-images.githubusercontent.com/18496702/174639745-81fbda3d-6e34-400f-8991-dc256c701eae.png)
![L11](https://user-images.githubusercontent.com/18496702/174639756-0a8ed930-0a7f-406e-a858-7eb03a295764.png)
L10p1 and L10p2 are two pages of part of the diagrams of Line10 process.
And L11 is part of the diagrams of Line11 process.
In L10p1, diagram 5 is overlapping diagram 1 in L11, so is diagram 7 in L10p2 with diagram 3 in L11.

Now I understand the logic. p p > ts wp j contains gluon radiation from p p process as well as radiation from the top.
Applying this logic, Line13 is overlapping with Line 12 as well, thus should be removed too.
But removing those process will at the same time omits the diagrams where the b radiates a gluon jet.

Thanks for pointing them out.

menglu21 replied:
Hi, could you please have a check on the difference between with and without line11. i assume line11 could be handled by the parton shower and maybe could be omitted

sifuluo replied:
With L11:
=== Results Summary for run: run_01 tag: tag_1 ===

     Cross-section :   0.2242 +- 0.001832 pb
     Nb of events :  1000
     Pythia8 merged cross-sections are:
      > Merging scale = 45     : 0.091709    +/- 0.00045 [pb]
      > Merging scale = 67.5   : 0.097315    +/- 0.00047 [pb]
      > Merging scale = 90     : 0.097539    +/- 0.00047 [pb]
    1200 diagrams (156 independent).

Without L11:
=== Results Summary for run: run_01 tag: tag_1 ===

     Cross-section :   0.2212 +- 0.002185 pb
     Nb of events :  1000
     Pythia8 merged cross-sections are:
      > Merging scale = 45     : 0.088919    +/- 0.00044 [pb]
      > Merging scale = 67.5   : 0.094449    +/- 0.00046 [pb]
      > Merging scale = 90     : 0.096219    +/- 0.00046 [pb]
    1136 diagrams (148 independent)

For L11 alone:
  === Results Summary for run: run_01 tag: tag_2 ===

     Cross-section :   0.001472 +- 1.441e-05 pb
     Nb of events :  1000
     Pythia8 merged cross-sections are:
      > Merging scale = 45     : 0.00042254  +/- 2.5e-06 [pb]
      > Merging scale = 67.5   : 0.00060952  +/- 3e-06   [pb]
      > Merging scale = 90     : 0.00077589  +/- 3.4e-06 [pb]
    64 diagrams (8 independent)
I don't know how MG is handling the sub processes. The x-sections of the latter two don't sum to the WithL11 process, while the numbers of diagrams and independent diagrams do.

menglu21 replied:
hi @sifuluo thanks for the check, the difference seems small. btw, do you really need do the parton emission at matrix element level, i think the parton shower should be enough. e.g., https://github.com/cms-sw/genproductions/pull/2807 doesn't require the ME jet, the same for the ttbar production.

sihyunjeon replied:
I am not sure if jet merging will be stable if you try to add all of these up separately in ME level.
1. parton ISR
2. top FSR
3. bottom FSR
The jets after the resonance decays (so W > j j) won't be used for jet merging at all - this is what I know

menglu21 replied:
yes, the max jet number will be used for matching is 1.

sifuluo replied:
Thanks for the tips and the script to generate djr plots.
We decided to remove all but the first two processes.
Now the x-section becomes
=== Results Summary for run: run_01 tag: tag_1 ===

     Cross-section :   0.2031 +- 0.001606 pb
     Nb of events :  1000
     Pythia8 merged cross-sections are:
      > Merging scale = 45     : 0.080824    +/- 0.0004  [pb]
      > Merging scale = 67.5   : 0.081636    +/- 0.00041 [pb]
      > Merging scale = 90     : 0.081027    +/- 0.00041 [pb]

944 diagrams (124 independent)

And the DJR plots are produced with nJetMax = 1, xqcut = 30 and qcut = 45
[DJR01.pdf](https://github.com/cms-sw/genproductions/files/8975463/DJR01.pdf)
[DJR12.pdf](https://github.com/cms-sw/genproductions/files/8975464/DJR12.pdf)
[DJR23.pdf](https://github.com/cms-sw/genproductions/files/8975465/DJR23.pdf)
[DJR34.pdf](https://github.com/cms-sw/genproductions/files/8975466/DJR34.pdf)

sifuluo replied:
For 10k events:
  === Results Summary for run: run_02 tag: tag_1 ===

     Cross-section :   0.2054 +- 0.000767 pb
     Nb of events :  10000
     Pythia8 merged cross-sections are:
      > Merging scale = 45     : 0.085621    +/- 0.00013 [pb]
      > Merging scale = 67.5   : 0.08901     +/- 0.00014 [pb]
      > Merging scale = 90     : 0.089544    +/- 0.00014 [pb]
[DJR01.pdf](https://github.com/cms-sw/genproductions/files/8976561/DJR01.pdf)
[DJR12.pdf](https://github.com/cms-sw/genproductions/files/8976562/DJR12.pdf)
[DJR23.pdf](https://github.com/cms-sw/genproductions/files/8976563/DJR23.pdf)
[DJR34.pdf](https://github.com/cms-sw/genproductions/files/8976564/DJR34.pdf)

sifuluo commented:
> As for the jet matching setup, you should set njetmax=1 since there is only 1 QCD jet which needs matching. for the djr, you are expected to check plots like https://deschwar.web.cern.ch/deschwar/tWZ/plots/DJR/DJR01_WZ.pdf to determine the xqcut and qcut

Is there standard process to produce such plots?

menglu21 commented:
> > As for the jet matching setup, you should set njetmax=1 since there is only 1 QCD jet which needs matching. for the djr, you are expected to check plots like https://deschwar.web.cern.ch/deschwar/tWZ/plots/DJR/DJR01_WZ.pdf to determine the xqcut and qcut
> 
> Is there standard process to produce such plots?

sorry, i overlooked the content of the plots, those plots are directly from MG. what value you used for "jetmatching:njetmax", the correct number should be 1.

menglu21 commented:
> > > As for the jet matching setup, you should set njetmax=1 since there is only 1 QCD jet which needs matching. for the djr, you are expected to check plots like https://deschwar.web.cern.ch/deschwar/tWZ/plots/DJR/DJR01_WZ.pdf to determine the xqcut and qcut
> > 
> > 
> > Is there standard process to produce such plots?
> 
> sorry, i overlooked the content of the plots, those plots are directly from MG. what value you used for "jetmatching:njetmax", the correct number should be 1.

btw, i put some code here to make the djr plots have customized style. https://github.com/cms-sw/genproductions/pull/3203

menglu21 commented:
> I produced DJR plots with the new process card and the script @menglu21 provided, with the file name formated as (xqcut - qcut.pdf) Note in the run_card setting for this iteration, I have set `lhapdf = pdlabel` `230000 = lhaid` as the card default. because `$DEFAULT_PDF_SETS = lhaid` will fail the process due to missing value for `$DEFAULT_PDF_SETS`
> 
> [30-45.pdf](https://github.com/cms-sw/genproductions/files/8993041/30-45.pdf) [30-60.pdf](https://github.com/cms-sw/genproductions/files/8993048/30-60.pdf) [60-90.pdf](https://github.com/cms-sw/genproductions/files/8993051/60-90.pdf) [10-15.pdf](https://github.com/cms-sw/genproductions/files/8993053/10-15.pdf) [10-30.pdf](https://github.com/cms-sw/genproductions/files/8993056/10-30.pdf)

the DJR plots look weird to me, there should be separation between different parton number (in official production there is truncation). i did a test on the DY: p p > e+ e- & p p > e+ e- j, with xqcut=10 and qcut=20, the djr plots[1] look reasonable to me. i would suggest you to do some further studies:
1. try other xqcut and qcut combination
2. try to generate gridpack and follow the official stream to generate some events and check the djr using https://github.com/cms-sw/genproductions/blob/master/bin/MadGraph5_aMCatNLO/macros/plotdjr.C

[1] https://melu.web.cern.ch/melu/DY_djr/

sifuluo commented:
We noticed this non-separation in our group meetings.
This is also the case in Andrew's pull request https://github.com/cms-sw/genproductions/pull/2807
And BFF ZPrime process, which can be found at the second item on the agenda https://indico.cern.ch/event/714080/#2-bff-zprime-follow-up

So we are not expecting to see a clear separation in 3rd generation and gluon fusion mixed processes.
"This is presumably because there's some ambiguity between the pdf definition and gluon splitting that's not entirely resolvable."  as our postdoc Denis Rathjens said about the case of BFF Z'.

sifuluo commented:
@menglu21 Are there any further issues or comments regarding my PR?

sihyunjeon commented:
Hi can you elaborate more on "This is presumably because there's some ambiguity between the pdf definition and gluon splitting that's not entirely resolvable."? At least for ttbar NLO FXFX [1] doesn't have such thing. I was the one who merged the PR #2807 but I think I missed that there was no separation in the plots (and I only realized it now) so I am not sure if that's really the physics effect or some problems with the module we have for DJR plots.

[1] https://indico.cern.ch/event/1068438/contributions/4565462/attachments/2329689/3969600/GEN_2021-10-18_jetmerging.pdf

sifuluo commented:
Let me quote an email from Michelangelo Mangano <michelangelo.mangano@cern.ch> discussing this behavior on Z' diagrams:

> the rule of thumb is that the xqcut should be in the range of alpha_s * sqrt(shat), so typically 10% of the energy of the hard process. Thus 30-50 GeV is ok as a range for a 500GeV Z'. The process you are studying is however rather peculiar, because of the initial state b's. The modeling of the initial state evolution for a b quark in the MCs is not greatly accurate, and this will be reflected in a poor matching of the samples with different multiplicity. 
>     For example, your 1-jet sample will likely be dominated by gb->bZ', rather than by bb->gZ'. The gb->bZ' process appears in the shower evolution of bb->Z', when the initial-state b evolves backward into a gluon that splits into b+bbar. But if the g->bb splitting is not well modeled, then you find a mismatch between the 0-jet plus shower and the 1-jet rate (you probably are aware of long-standing issues on the 4- vs 5-flavour approach to calculations involving initial-state heavy flavours, and of the systematic differences between the two approaches, unless specific scale choices are made ... what you are doing will certianly expose these issues!)
>     The changes at the level of 20% that you report, therefore, seem to me totally acceptable, considering that this is a LO process, and considering the above remark about heavy quarks.

Hope this can explain

sihyunjeon commented:
How does this explain the issue with nonresolving DJR plots?

sifuluo commented:
Thanks for getting back to me.
We are convinced that the b-quark is the cause for this effect.
And we will discuss this issue and try to get more understanding of it.

sihyunjeon commented:
I don't see a point with b quarks. b quark having not so good modelling means that the calculations are not trustable and that's it. What we see is a DJR plot that splits the sample into subset : "N-jet samples" and plotting them. How is b quark only problematic and not for other quarks?

sifuluo commented:
Hi @sihyunjeon @menglu21 ,
In order to check the behavior, I checked some process.
I first tried to reproduce the DY process as [1] and the djr looks like [2], same as what Menglu has produced.
So our workflow should be alright.

Then I limited the process to coupling only to b and g as [3] and the result looks like [4].
As you said, b quark alone did not explain the overlap.

I then checked the mass dependence of the djr overlap, with m_{W'} = [300,500,1800].
The results are shown in [5][6][7].
This set of plots exhibits a mass dependent distribution. And lower mass seems to be more separated.
And higher mass shows an enhancement to 1 parton distribution.
To my understanding, as the resonance mass increases, the b-pdf will contribute less, and the gluon splitting will take over more proportion, that leads to the a higher 1 parton distribution. And this enhancement in 1 parton distribution might be the cause of the overlap.

To confirm this overlap will shrink in the low mass region, I further tested a m_{W'} = 80, which is roughly the SM W mass.
Since with such a low mass, it cannot decay to an on-shell top quark, I tested with a process in [8].
The results are shown in [9], which is similar to DY results and showing a clear separation.

So to our understanding, the overlap is reasonable.

[1]
```
define p = p b b~
define j = j b b~
define ls = mu+ mu- e+ e-
generate p p > ls ls 
add process p p > ls ls j
add process p p > ls ls j j
```

[2] [ee.pdf](https://github.com/cms-sw/genproductions/files/9061623/ee.pdf)

[3]
```
define j = j b b~
define ls = mu+ mu- e+ e-
generate b b~ > ls ls 
add process g b~ > ls ls b~
add process b g > ls ls b
add process g g > ls ls b b~
```

[4] [eeb.pdf](https://github.com/cms-sw/genproductions/files/9061644/eeb.pdf)

[5] [FL300.pdf](https://github.com/cms-sw/genproductions/files/9061656/FL300.pdf)

[6] [FL500.pdf](https://github.com/cms-sw/genproductions/files/9061659/FL500.pdf)

[7] [FL1800.pdf](https://github.com/cms-sw/genproductions/files/9061661/FL1800.pdf)

[8]
```
import model WPrimetb_5f_UFO
define tas = ta+ ta-
define nuta = vt vt~
define wp = wp+ wp-
generate p p > wp, wp > tas nuta
add process p p > wp j, wp > tas nuta
```
[9] [lowwp.pdf](https://github.com/cms-sw/genproductions/files/9061752/lowwp.pdf)

sihyunjeon commented:
Let's clarify my question.

<img width="961" alt="image" src="https://user-images.githubusercontent.com/17442656/177923789-bb38c2eb-b0fd-4921-81ac-0944ddf0644a.png">

This is from my [slide](https://indico.cern.ch/event/1068438/contributions/4565462/attachments/2329689/3969600/GEN_2021-10-18_jetmerging.pdf) page9 on the right. DJR is the measure of separation between matrix element level and parton shower jet. So the reason for the cutoff at DJR=1.3 there is because qcut (jet merging scale) is set to 22 and log10(22) ~ 1.3. What is the reason for your plots not having cutoffs at all, was my first question. So in your last reply,
> And lower mass seems to be more separated.

I don't see what you mean. Nothing is separated and DJR cutoff (log10(qcut)) which should be shown in the plots are nowhere for all the mass cases.

sifuluo commented:
I am sorry that I am confused by the definition or mechanism of DJR plots production. We don't have an expert that we can contact to ask for details. For us, it is like a black box, and we don't know what to expect in the first place.
I worked this way because I didn't see such cutoffs in Menglu's DY DJR[1] either.
Why don't DY process have such clear cutoffs?

[1]https://melu.web.cern.ch/melu/DY_djr/

sifuluo commented:
Hi @sihyunjeon 
Are you able to figure out what is wrong with @menglu21 's DJR plots, that there isn't a clear cut off in DY as you presented?
I think that could be the key for the absence of cut-off in my case as well.

sifuluo commented:
Hi @sihyunjeon @menglu21 
With the help from @menglu21 ,I manged to reproduce the DY DJR plots with the steep cut-off [1].
I only did with p p > e e and p p > e e j. So there is a cliff on the right side of the plot DJR 1->2.

I then ran over my W' model also with only 1 additional jet and obtained DJR plot as [2].
The cliff in DJR 1->2 plot persists without surprise. And also a jump in DJR 0->1 plot is observed.
So I will try higher qcut value to smooth the jump.
But does the cliff in DJR 1->2 plot matter? And should I generate process with higher multiplicity of additional jets to complete for the cliff?

[1](https://github.com/cms-sw/genproductions/files/9202464/DY.pdf)
[2](https://github.com/cms-sw/genproductions/files/9202489/wpdjr.pdf)

sifuluo commented:
Thanks for the reply.
The CMSSW pythia fragment is the same with Andrew's
https://github.com/cms-sw/genproductions/blob/master/genfragments/ThirteenTeV/Hadronizer/Hadronizer_TuneCP5_13TeV_MLM_5f_max2j_qCut45_LHE_pythia8_cff.py
With cmsDriver command, I get a cmsRun python script.
Then I changed the gridpack path and added some lines in the head to make changing easier:
https://gist.github.com/sifuluo/38bddbd575d07e95974a2c91b1ce9064
That is the code I cmsRun with.

As for the standalone Madgraph Pythia8 card:
https://gist.github.com/sifuluo/269d1ff27fcd1be88e5fd6783c2851a3

sihyunjeon commented:
'JetMatching:nJetMax = 2', #number of partons in born matrix element for highest multiplicity

This should be changed to 1 and this is the reason for the strange cutoff at 1->2

sifuluo commented:
Right, that makes perfect sense. I will run it again but with a higher qcut value to see if the all parton distribution looks smooth.

sifuluo commented:
Have you figured out why standalone madgraph produced DJR plots don't have cut offs?

sihyunjeon commented:
FYI, that line expects you to have 2 additional partons,
p p > X
p p > X j
p p > X j j

But you only have
p p > X
p p > X j
So should be 1

sihyunjeon commented:
> Have you figured out why standalone madgraph produced DJR plots don't have cut offs?

I never really used MadGraph standalone myself - but at least CMS MC sample uses GEN datatier to do DJR studies for SM processes so I would rather believe CMS ones

sifuluo commented:
Hi,
I think xqcut = 30 qcut = 80 is yielding a rather smooth DJR distribution:
[CMSSW30-80.pdf](https://github.com/cms-sw/genproductions/files/9232075/CMSSW30-80.pdf)
Should I add the hadronizer in this PR or in a separate one?

sifuluo commented:
Hi @sihyunjeon  @menglu21 
I further tested on mass = 300 case with different xqcuts/qcuts:
Among them, xqcut = 30 qcut = 80 seems to be also the best choice.
[M300_20-60.pdf](https://github.com/cms-sw/genproductions/files/9251242/M300_20-60.pdf)
[M300_30-45.pdf](https://github.com/cms-sw/genproductions/files/9251244/M300_30-45.pdf)
[M300_30-60.pdf](https://github.com/cms-sw/genproductions/files/9251245/M300_30-60.pdf)
[M300_30-80.pdf](https://github.com/cms-sw/genproductions/files/9251247/M300_30-80.pdf)

Are there any further tests you would like to see before it can be merged?

sihyunjeon commented:
hi it looks good

sifuluo commented:
Hi @sihyunjeon, I just updated the hadronizer for its production.

I think it is ready to be merged.

sifuluo commented:
Hi @sihyunjeon @menglu21 ,
Could you please review my PR and approve it or pin someone that can?
Or if there is anything subject to change, please let me know.

sihyunjeon commented:
@mlizzo @menglu21 @Saptaparna @GurpreetSinghChahal  ping

@sifuluo i am not the person who make merges anymore :)

sifuluo commented:
@sihyunjeon  Thanks!

mlizzo commented:
@sifuluo @sihyunjeon , sorry I was on vacation until monday. It seems to me this PR has been extensively reviewed already and I don't have further comments, I'll merge it immediately.

sifuluo commented:
@mlizzo Thanks a lot!
